packages/@react-spectrum/tooltip/src/TooltipTrigger.tsx

```diff
@@ -10,78 +10,73 @@
  * governing permissions and limitations under the License.
  */
 
-import {DOMPropsResponder, PressResponder} from '@react-aria/interactions';
+import {DOMPropsResponder} from '@react-aria/interactions';
 import {Overlay} from '@react-spectrum/overlays';
-import React, {Fragment, useRef} from 'react';
+import {PlacementAxis} from '@react-types/overlays';
+import React, {Fragment, RefObject, useContext, useRef} from 'react';
+import {StyleProps} from '@react-types/shared';
 import {TooltipTriggerProps} from '@react-types/tooltip';
 import {useOverlayPosition} from '@react-aria/overlays';
 import {useTooltipTrigger} from '@react-aria/tooltip';
 import {useTooltipTriggerState} from '@react-stately/tooltip';
 
+interface TooltipContextProps extends StyleProps {
+  overlayRef?: RefObject<HTMLDivElement>,
+  placement?: PlacementAxis,
+  isOpen?: boolean
+}
+
+const TooltipContext = React.createContext<TooltipContextProps>({});
+
+export function useTooltipProvider(): TooltipContextProps {
+  return useContext(TooltipContext);
+}
+
 export function TooltipTrigger(props: TooltipTriggerProps) {
   let {
     children,
-    type,
-    targetRef,
-    isOpen,
     isDisabled
   } = props;
 
-  let [trigger, content] = React.Children.toArray(children);
+  let [trigger, tooltip] = React.Children.toArray(children);
 
-  let state = useTooltipTriggerState(props);
+  let state = useTooltipTriggerState();
 
   let triggerRef = useRef<HTMLElement>();
   let overlayRef = useRef<HTMLDivElement>();
 
   let {triggerProps, tooltipProps} = useTooltipTrigger({
-    tooltipProps: content.props,
-    triggerProps: {
-      ...trigger.props,
-      ref: triggerRef
-    },
-    state,
-    isDisabled,
-    type
-  });
+    tooltipProps: tooltip.props,
+    triggerProps: trigger.props,
+    isDisabled
+  }, state, triggerRef);
 
-  let {overlayProps, placement, arrowProps} = useOverlayPosition({
+  let {overlayProps, placement} = useOverlayPosition({
     placement: props.placement,
-    targetRef: targetRef || triggerRef,
-    overlayRef,
-    isOpen
+    targetRef: triggerRef,
+    overlayRef
   });
 
-  let overlay = (
-    <Overlay isOpen={state.open}>
-      {React.cloneElement(content, {placement: placement, arrowProps: arrowProps, ref: overlayRef, UNSAFE_style: overlayProps.style, isOpen: open, ...tooltipProps})}
-    </Overlay>
+  return (
+    <Fragment>
+      <DOMPropsResponder
+        {...triggerProps}
+        ref={triggerRef}
+        isDisabled={isDisabled}>
+        {trigger}
+      </DOMPropsResponder>
+      <Overlay isOpen={state.open}>
+        <TooltipContext.Provider
+          value={{
+            placement,
+            overlayRef,
+            UNSAFE_style: overlayProps.style,
+            isOpen: state.open,
+            ...tooltipProps
+          }}>
```

Comment:
Hrm, digging into this currently but it seems like the diagonal tooltip sliding that happens on first render is due to the change from using cloneElement to using this context, maybe something with how the refs?

Reply:
After digging into this some more, it seems the diagonal sliding is happening because the tooltip is rendered out of position initially and `isOpen` is true while it is still trying to get the tooltip to the right place. I got the sliding to go away by appending doing a pretty silly conditional `'is-open': isOpen && styleProps.style.maxHeight` in `Tooltip.tsx` but I haven't dug into why isOpen is false initially for the original code

Reply:
Ah, its because of how `Overlay`-> `OpenTransition`->`Transition` provides a `isOpen` prop to its children via cloneElement. To fix I think you'll have to change the ordering here so the `TooltipContext.Provider` wraps `Overlay` instead, then you'll have drop the `isOpen` you pass via the TooltipProvider here so it doesn't overwrite the `props.isOpen` in Tooltip provided by the cloneElement that happens in `OpenTransition`.

```
        <TooltipContext.Provider
          value={{
            placement,
            overlayRef,
            UNSAFE_style: overlayProps.style,
            ...tooltipProps
          }}>
            <Overlay isOpen={state.open}>
              {tooltip}
            </Overlay>
        </TooltipContext.Provider>
```

```

```

Reply:
thnaks! that was very helpful :) reordered provider without isOpen and it works well https://github.com/adobe-private/react-spectrum-v3/pull/426/commits/22f7fff1a6ef53e94c7c9b051c62ee2ed5733e6e